### PR TITLE
feat(manager:conda): add manager for conda

### PR DESCRIPTION
**Pull request by morremeyer**
## Changes

This is a first implementation of a manager for conda as discussed in #2213.

It has very limited support for version specifications, namely: only exact matching of release versions that have three components and can therefore be assumed to be semver compliant.

## Context

Conda [versioning](https://github.com/conda/conda/blob/0ce67ff01354667d6cf8c956839a401f42e41bde/conda/models/version.py#L49) and [dependency definitions](https://github.com/conda/conda/blob/0ce67ff01354667d6cf8c956839a401f42e41bde/conda/models/match_spec.py#L69) are very different from what renovate already supports. 

@ekeih and I discussed how to best get something of value to all renovate users where we can guarantee that updates are correct, but without making this a huge initial implementation with full conda versioning support and all shenanigans.

We therefore implemented this limited support to get started and get familiar with the codebase (again).

If you're visiting this PR in the future: We're actively talking to the maintainers about how we can improve the support over time. There's already several ideas for improvements already.

## Documentation

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/conda/extract.spec.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  describe('extractPackageFile', () => {
+    it('returns null for empty', () => {
+      expect(extractPackageFile('not an environment.yml')).toBeNull();
+    });
+
+    it('extracts dependencies', () => {
+      const res = extractPackageFile(environmentYml);
+      expect(res?.deps.filter((dep) => dep.skipReason)).toHaveLength(9);
+      expect(res?.deps).toHaveLength(12);
+      expect(res).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/manager/conda/extract.spec.ts:16]: please read our best practices guide and don't use toEqual or toMatchObject. reduce deps in fixture to minimum 

guide is in docs/development folder

File: lib/modules/manager/conda/extract.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+// This currently only supports strict semver versions. conda also supports
+// versions with 4 numeric components (e.g. 7.10.3.1)
+// As we do only support exact matching with strict semver as of now, this
+// regex is sufficient for now.
+const releaseVersionRegex = regEx(/^(?<version>(\d+\.){1,2}\d)$/);
+
+export function extractPackageFile(content: string): PackageFile | null {
+  let parsedContent: CondaEnvironment;
+  try {
+    parsedContent = load(content, { json: true }) as any;
```
Comment (viceice) [inline lib/modules/manager/conda/extract.ts:21]: cast to correct type

File: lib/modules/manager/conda/extract.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+);
+
+// This currently only supports strict semver versions. conda also supports
+// versions with 4 numeric components (e.g. 7.10.3.1)
+// As we do only support exact matching with strict semver as of now, this
+// regex is sufficient for now.
+const releaseVersionRegex = regEx(/^(?<version>(\d+\.){1,2}\d)$/);
+
+export function extractPackageFile(content: string): PackageFile | null {
+  let parsedContent: CondaEnvironment;
```
Comment (viceice) [inline lib/modules/manager/conda/extract.ts:19]:
```suggestion
  let parsedContent: CondaEnvironment | null;
```
yaml load can return null for empty docs

File: lib/modules/manager/conda/extract.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    // This is only done to prevent unnecessary HTTP requests
+    if (channels.length === 0) {
+      return {
+        deps: [
+          {
+            depName: 'nodefaults',
+            skipReason: 'invalid-config',
+          },
+        ],
+      };
```
Comment (viceice) [inline lib/modules/manager/conda/extract.ts:51]: return null instead

File: lib/modules/manager/conda/extract.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        return obj !== 'defaults';
+      });
+    }
+
+    // Always add default channels
+    channels.push('main', 'r');
+  }
+
+  const deps: PackageDependency[] = [];
+  parsedContent.dependencies?.forEach((parsedDep) => {
```
Comment (viceice) [inline lib/modules/manager/conda/extract.ts:70]: use `for...of` loop, checkout best practices

File: lib/modules/manager/conda/extract.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+  if (!dep.skipReason && !dep.packageName) {
+    const channelDeps: PackageDependency[] = [];
+    for (const channel of channels) {
+      // This is a shallow copy! As we can keep all references that this shallow copy holds,
+      // this is okay. If you read this comment and have a better idea, please let us know!
+      const cloneDep = Object.assign({}, dep);
+
+      /* eslint-disable-next-line @typescript-eslint/restrict-template-expressions */
+      cloneDep.depName = `${channel}/${cloneDep.depName}`;
+      cloneDep.packageName = cloneDep.depName;
```
Comment (viceice) [inline lib/modules/manager/conda/extract.ts:161]: if it's same as depName, remove

File: lib/modules/manager/conda/extract.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+  }
+
+  // Check if the version is strict semver
+  const versionMatch = releaseVersionRegex.exec(groups.version);
+
+  // If exact matching is used (name==version), we know the exact
+  // version used and set it explicitly.
+  if (versionMatch?.groups?.version && groups.exact === '==') {
+    dep.currentVersion = groups.version;
+    dep.versioning = 'semver';
```
Comment (viceice) [inline lib/modules/manager/conda/extract.ts:128]: i think you should use pep440 versioning instead, as it will support most of your limitations

**Review by viceice — COMMENTED** [review]
it seems channels should be part of the registry url, one for each channel. that would make more sense in my eyes

File: lib/modules/manager/conda/readme.md [hunk 8]
```diff
@@ -0,0 +1,17 @@
+Supports conda `environment.yml` files.
+
+Renovate has no concept of conda channels and their different priorities. The manager will open separate PRs for updates from different channels. This allows you to make a decision about which updates to merge.
```
Comment (HonkingGoose) [inline lib/modules/manager/conda/readme.md:3]:
```suggestion
Renovate has no concept of conda channels and their different priorities.
The manager will open separate PRs for updates from different channels.
This allows you to make a decision about which updates to merge.
```

We use the "one sentence per line" pattern in our docs.

File: lib/modules/manager/conda/readme.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+**Warning**
+
+When enabling the manager, you will get _Package Lookup Warnings_ for all packages that are not present in all channels. This is, again, due to renovate not having a concept of conda channels and treating each package for each channel as a separate dependency. You can safely ignore those warnings for a package if they do not occur for all channels you are using (including the default channels).
```
Comment (HonkingGoose) [inline lib/modules/manager/conda/readme.md:7]:
```suggestion
When enabling the manager, you will get _Package Lookup Warnings_ for all packages that are not present in all channels.
This is, again, due to Renovate not having a concept of conda channels and treating each package for each channel as a separate dependency.
You can safely ignore those warnings for a package if they do not occur for all channels you are using (including the default channels).
```

File: lib/modules/manager/conda/readme.md [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+**Limitations**
+
+The `conda` manager only supports a subset of the conda environment.yml specification. The following limitations exist:
```
Comment (HonkingGoose) [inline lib/modules/manager/conda/readme.md:11]:
```suggestion
The `conda` manager only supports a subset of the conda `environment.yml` specification.
The following limitations exist:
```

File: lib/modules/manager/conda/readme.md [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+- Only [release versions](https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/pkg-specs.html#supported-version-strings) that are strictly semantically versioned are supported (e.g. `2.1.5`, but not `3.1.7.6`).
+- The default `msys2` channel for Windows is not added when using default channels as renovate does not know which operating system you will use the `environment.yml` file with. If you use Windows, you will have to explicitly include the `msys2` channel to get all updates.
+- key-value arguments in dependency specifications are not supported.
+- The key `pip` in the `dependencies` list is ignored if it is a list.
+- There is no support for the nonstandard keys introduced by [conda-lock](https://github.com/conda-incubator/conda-lock).
```
Comment (HonkingGoose) [inline lib/modules/manager/conda/readme.md:17]:
```suggestion
- Only [release versions](https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/pkg-specs.html#supported-version-strings) that are strictly semantically versioned are supported. For example: `2.1.5` is supported but `3.1.7.6` is not
- The default `msys2` channel for Windows is not added when using default channels as Renovate doesn't know which operating system you will use the `environment.yml` file with. If you use Windows, you will have to explicitly include the `msys2` channel to get all updates
- key-value arguments in dependency specifications are not supported
- The key `pip` in the `dependencies` list is ignored if it is a list
- Renovate doesn't support the non-standard keys introduced by [conda-lock](https://github.com/conda-incubator/conda-lock)
```

Changes:

- Drop `.` at the end of list items to match the style in the rest of the docs
- Rewrite some sentences
- Small grammar/typo fixes

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
@morremeyer let's close this after 6 months inactivity. Of course, hopeful you can reopen or replace it soon :)

**Comment by morremeyer** [comment]
@rarkins Good call, yes. I hope we can get to work on this again soon.

We definitely already have a much better idea of how to go about implementing it better than this first attempt.

**Pull request closed without merging**